### ISSUE TITLE
help: go-git failed to checkout with tag and unstaged changes

### ISSUE THREAD

**Issue by khareyash05** [fluxcd/flux2#5027]
### Describe the bug

To set the context, we have a flux cd setup 
```
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: GitRepository
metadata:
  name: api-server-staging
  namespace: flux-system
spec:
  interval: 1m
  url: ssh://git@github.com/......
  ref:
    branch: main
---
```
Since flux uses go-git, we are facing issue of 
`failed to checkout and determine revision: unable to checkout tag 'vX.X.X': worktree contains unstaged changes`

We did a backtrack of where this issue is and found it in https://github.com/fluxcd/pkg/blob/ac1007b57e37838e73b8bc95365dab9a0e856e8e/git/gogit/clone.go#L382

We then isolated it and created our own program simulating the same, and were able to replicate the issue.It looks like that go-git is able to checkout the code but not able to go to that specific tag because it does a sort of  `soft reset` ie. if we checkout 0.7.3 all the future commits to the branch move to the staging area and thus the former issue of `unstaged changes`

We want to create a Helm Chart citing the latest release , but since it fails at this step, we arent able to proceed further

### Steps to reproduce

Configure FluxCD for your environment

### Expected behavior

Cloning is done with working tree clean

### OS / Distro

Ubuntu

### Flux version

Latest

### Flux check

NA

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by khareyash05** [comment]
Also we did some diving into the code and found that if we use `Force:true` in https://github.com/fluxcd/pkg/blob/ac1007b57e37838e73b8bc95365dab9a0e856e8e/git/gogit/clone.go#L378C2-L380C4 , we were able to clone perfectly or if we put `NoCheckout:true` in https://github.com/fluxcd/pkg/blob/ac1007b57e37838e73b8bc95365dab9a0e856e8e/git/gogit/clone.go#L287C2-L298C3 , all work perfectly done, but not sure how we will do this using the flux config yaml

**Comment by khareyash05** [comment]
```
kubectl describe gitrepository -n flux-system api-server
Name:         api-server
Namespace:    flux-system
Labels:       kustomize.toolkit.fluxcd.io/name=flux-system
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  source.toolkit.fluxcd.io/v1beta2
Kind:         GitRepository
Metadata:
  Creation Timestamp:  2024-10-09T07:47:13Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        1
  Resource Version:  473625663
  UID:               a85c9410-3aaf-48e0-a796-e6017cc5843f
Spec:
  Git Implementation:  go-git
  Ignore:              # exclude all
/*
# include charts directory
!/deployment/

  Interval:  1m
  Ref:
    Semver:  >=0.1.0
  Secret Ref:
    Name:   api-server-github-ssh-key
  Timeout:  60s
  URL:      ssh://git@github.com/keploy/api-server
Status:
  Conditions:
    Last Transition Time:  2024-10-09T14:06:20Z
    Message:               building artifact
    Observed Generation:   1
    Reason:                ProgressingWithRetry
    Status:                True
    Type:                  Reconciling
    Last Transition Time:  2024-10-09T14:06:20Z
    Message:               failed to checkout and determine revision: unable to checkout tag 'v0.7.3': worktree contains unstaged changes
    Observed Generation:   1
    Reason:                GitOperationFailed
    Status:                False
    Type:                  Ready
    Last Transition Time:  2024-10-09T10:49:40Z
    Message:               failed to checkout and determine revision: unable to checkout tag 'v0.7.3': worktree contains unstaged changes
    Observed Generation:   1
    Reason:                GitOperationFailed
    Status:                True
    Type:                  FetchFailed
  Observed Generation:     -1
Events:
  Type     Reason              Age                   From               Message
  ----     ------              ----                  ----               -------
  Warning  GitOperationFailed  13m (x27 over 6h13m)  source-controller  failed to checkout and determine revision: unable to checkout tag 'v0.7.3': worktree contains unstaged changes
```

**Comment by gecube** [comment]
**Triage: reproduced end-to-end on the current stack (source-controller v1.8.5, fluxcd/pkg git v0.52.0, go-git v5.19.1). Root cause identified, one-line fix validated locally.**

## TL;DR

`spec.ref.semver` is the only ref type that makes the gogit client perform a **second, non-forced checkout** after the clone ([`cloneSemVer`](https://github.com/fluxcd/pkg/blob/git/v0.52.0/git/gogit/clone.go#L417-L423)). If go-git's `Worktree.Status()` produces a false positive on the freshly-cloned worktree, that checkout aborts with `worktree contains unstaged changes`. One confirmed trigger for such a false positive is an **orphan gitlink** — a nested git repository accidentally committed (index entry mode `160000`) without a matching `.gitmodules` entry. The very same repo and tag work fine via `ref.tag` or `ref.branch`, and stock `git checkout v0.7.3` on the same clone is clean.

## Reproduction

Ready-made repro repo: **https://github.com/gecube/flux5027-repro** — it contains the orphan gitlink and tags `v0.1.0`/`v0.7.3`, plus ready-to-apply `GitRepository` manifests and a standalone Go client (`repro/`) that drives fluxcd/pkg exactly the way source-controller does.

To build such a repo from scratch:

```sh
git init app && cd app
echo hello > app.txt
git init vendored && (cd vendored && echo x > f.txt && git add -A && git commit -m inner)
git add app.txt vendored   # git prints: "hint: You've added another git repository inside your current repository"
git commit -m "app + orphan gitlink"
git tag v0.7.3
git push origin main --tags
```

`git ls-files -s` shows the gitlink, and there is no `.gitmodules`:

```
100644 ce013625... 0	app.txt
160000 d71f3a47... 0	vendored
```

Point three GitRepositories at it (source-controller v1.8.5):

| `spec.ref` | Result |
|---|---|
| `semver: ">=0.1.0"` | ❌ `failed to checkout and determine revision: unable to checkout tag 'v0.7.3': worktree contains unstaged changes` |
| `tag: v0.7.3` | ✅ `stored artifact for revision 'v0.7.3@sha1:255a28d…'` |
| `branch: main` | ✅ |

The error is verbatim the one reported here. It reproduces identically with the library versions Flux shipped at the time of this report (pkg/git/gogit v0.21.0 + go-git v5.12.0) and with today's fluxcd/pkg git v0.52.0 + go-git v5.19.1, in shallow and full clone modes alike. It fails even when the resolved tag **is** the default branch HEAD. Canonical git on the same repository checks the tag out cleanly (`git status --porcelain` is empty).

## Root cause

1. For `semver`, the client clones the default branch (checking out its HEAD), resolves the winning tag, and then switches to it with a non-forced checkout: [clone.go#L417-L423](https://github.com/fluxcd/pkg/blob/git/v0.52.0/git/gogit/clone.go#L417-L423).
2. In go-git, a non-forced `Checkout` runs `Reset(MergeReset)`, which calls [`containsUnstagedChanges()`](https://github.com/go-git/go-git/blob/v5.19.1/worktree.go#L517) → `Status()`. The filesystem side of the status diff only synthesizes nodes for submodules **listed in `.gitmodules`** ([`getSubmodulesStatus`](https://github.com/go-git/go-git/blob/v5.19.1/worktree_status.go#L202-L225) feeding [`filesystem.NewRootNodeWithOptions`](https://github.com/go-git/go-git/blob/v5.19.1/worktree_status.go#L139-L145)). A `160000` index entry that is *not* in `.gitmodules` has no filesystem counterpart after a fresh clone (the directory is never created), so the merkletrie diff reports it as deleted → the worktree is considered dirty → `ErrUnstagedChanges`. Canonical git treats this state as clean.
3. The other ref types don't hit this: `branch` and `tag` clone the target ref directly (no second checkout), and the `commit` path has done its checkout with [`Force: true`](https://github.com/fluxcd/pkg/blob/git/v0.52.0/git/gogit/clone.go#L292-L296) ever since the gogit implementation was introduced. Only `semver` was left behind.
4. This also explains the "all the future commits to the branch move to the staging area" observation above: go-git's `Checkout` moves HEAD **before** performing the reset ([worktree.go#L164-L201](https://github.com/go-git/go-git/blob/v5.19.1/worktree.go#L164-L201)), so when the reset aborts, HEAD already points at the tag while the index and worktree still hold the old branch state — `git status` then renders the difference as staged changes.

## What it is not

Negative controls that do **not** reproduce (tested on both go-git v5.12.0 and v5.19.1): committed symlinks, proper submodules with `.gitmodules`, CRLF blobs + `* text=auto`, tracked-but-gitignored files, executable bits, empty files, unicode filenames. There may be other `Status()` false positives of this class, but the orphan gitlink reproduces the reported symptom exactly.

I couldn't inspect `keploy/api-server` itself (it's no longer public), so I can't confirm which false positive that particular repo hit. @khareyash05 if you still have access: `git ls-files -s | grep ^160000` plus a missing/incomplete `.gitmodules` would confirm it's the same trigger.

## Suggested fix

Add `Force: true` to the checkout in `cloneSemVer`, mirroring what `cloneCommit` already does — the worktree at this point is a fresh clone, there is nothing user-made to lose:

```go
err = w.Checkout(&extgogit.CheckoutOptions{
	Branch:                    tagRef.Name(),
	Force:                     true,
	SparseCheckoutDirectories: opts.SparseCheckoutDirectories,
})
```

Validated against fluxcd/pkg git v0.52.0: the failing case above turns green and returns the same revision as `ref.tag`, and all negative controls stay green. An alternative would be `NoCheckout: true` on the initial clone plus a single forced checkout of the resolved tag, which would also avoid materializing the default-branch worktree only to throw it away. The underlying `Status()` false positive arguably deserves an upstream go-git issue too (a related `.gitmodules`/gitlink divergence is go-git/go-git#2255).

I'm happy to open a PR to fluxcd/pkg with the one-liner and a regression test if the maintainers agree with the approach.

## Workarounds meanwhile

- pin the version with `ref.tag` instead of `ref.semver`, or
- clean up the stray gitlink in the source repo: `git rm --cached <path> && git commit` (or turn it into a proper submodule with a `.gitmodules` entry).

cc @matheuscscp
